app/wasm/test/custom_msg_test.go

```diff
@@ -209,6 +209,163 @@ func TestSwapMsg(t *testing.T) {
 				sdk.NewInt64Coin("uosmo", 2000000),
 			},
 		},
+
+		"exact in: 2 step multi-hop": {
+			msg: func(state BaseState) *wasmbindings.SwapMsg {
+				return &wasmbindings.SwapMsg{
+					First: wasmbindings.Swap{
+						PoolId:   state.StarPool,
+						DenomIn:  "ustar",
+						DenomOut: "uosmo",
+					},
+					// Note: you must use empty array, not nil, for valid Rust JSON
+					Route: []wasmbindings.Step{{
+						PoolId:   state.AtomPool,
+						DenomOut: "uatom",
+					}},
+					Amount: wasmbindings.SwapAmountWithLimit{
+						ExactIn: &wasmbindings.ExactIn{
+							Input:     sdk.NewInt(240000000),
+							MinOutput: sdk.NewInt(1999000),
+						},
+					},
+				}
+			},
+			initFunds: sdk.NewInt64Coin("ustar", 240000000),
+			finalFunds: []sdk.Coin{
+				// 240 STAR -> 6 OSMO
+				// 6 OSMO -> 2 ATOM (with minor rounding)
```

Comment:
It's clear to me now: The exact in / out sequence always define the same. input -> intermediate outputs -> final output.

And we handle the differences between them under the hood. I agree this is a better design / API than the Osmosis one. Was just confused / doubting about the exact meaning of the difference between them.